### ISSUE TITLE
Memory issues

### ISSUE THREAD

**Issue by lwoldt** [secile/UsbCamera#27]
Hello, when using an external video capture card, you may notice that the memory usage continues to increase, ultimately causing the program to crash. Is there a button for real-time release

**Comment by secile** [comment]
Hello.  
I think my library do not have problem of leaking memory.

1. library include 2 sample project, 'UsbCameraForms' and 'UsbCameraWpf'. Use these sample projects and let me know crash happens or not.
2. if possible, show me your code.
3. tell me product name of external video capture card.

**Comment by SankarGaneshKumar** [comment]
Im aslo facing the same issue. GC is not collecting. Memory keeps on increasing when it reaches 3.9 GB it ends in Out of memory exception. My camera is See3CAM_CU55 by eCam and its details are 5 MP (2592 x 1944) 24 fps and 12 fps Uncompressed
UYVY and Compressed MJPEG  60 fps . 
![Screenshot (33)](https://github.com/secile/UsbCamera/assets/18550832/9eced157-196e-4e0a-b99c-ed0ff6288de7)
![Screenshot (32)](https://github.com/secile/UsbCamera/assets/18550832/447a738b-a24b-46f7-aac2-7c7b9a9c0777)

**Comment by yangjieshao** [comment]
> Im aslo facing the same issue. GC is not collecting. Memory keeps on increasing when it reaches 3.9 GB it ends in Out of memory exception. My camera is See3CAM_CU55 by eCam and its details are 5 MP (2592 x 1944) 24 fps and 12 fps Uncompressed UYVY and Compressed MJPEG 60 fps . ![Screenshot (33)](https://user-images.githubusercontent.com/18550832/274487170-9eced157-196e-4e0a-b99c-ed0ff6288de7.png) ![Screenshot (32)](https://user-images.githubusercontent.com/18550832/274487176-447a738b-a24b-46f7-aac2-7c7b9a9c0777.png)

change
```
var timer = new System.Timers.Timer(1000 / 30) { SynchronizingObject = this };
timer.Elapsed += (s, ev) => pictureBox1.Image = camera.GetBitmap();
timer.start();
```
to
```
var timer = new System.Timers.Timer(1000 / 30) { SynchronizingObject = this };
timer.Elapsed += (s, ev) =>
{
	var oldImage = pictureBox1.Image;
	pictureBox1.Image = camera.GetBitmap();
	if(oldImage!=null)
	{
		oldImage.Dispose();
	}
};
timer.start();
```

**Comment by lwoldt** [comment]
改用其他库，该库存在明显的内存回收问题，起因是外接的usb设备无法管理



---Original---
From: ***@***.***&gt;
Date: Thu, Oct 12, 2023 19:30 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [secile/UsbCamera] Memory issues (Issue #27)




  
Im aslo facing the same issue. GC is not collecting. Memory keeps on increasing when it reaches 3.9 GB it ends in Out of memory exception. My camera is See3CAM_CU55 by eCam and its details are 5 MP (2592 x 1944) 24 fps and 12 fps Uncompressed UYVY and Compressed MJPEG 60 fps .  
  
change
 var timer = new System.Timers.Timer(1000 / 30) { SynchronizingObject = this }; timer.Elapsed += (s, ev) =&gt; pictureBox1.Image = camera.GetBitmap(); timer.start();  
to
 var timer = new System.Timers.Timer(1000 / 30) { SynchronizingObject = this }; timer.Elapsed += (s, ev) =&gt; { 	var oldImage = pictureBox1.Image; 	pictureBox1.Image = camera.GetBitmap(); 	if(oldImage!=null) 	{ 		oldImage.Dispose(); 	} }; timer.start();

**Comment by yangjieshao** [comment]
> 改用其他库，该库存在明显的内存回收问题，起因是外接的usb设备无法管理

并没有问题哦  我有项目跑了2年了 前台登记的客户端 一点内存都没漏
只是例子里  没释放pictureBox1.Image旧图片

你用啥库 只要是把图片clone一份到pictureBox1.Image 都得自己释放

**Comment by lwoldt** [comment]
您的库在系统自带的摄像头下无任何问题，一旦有其他的外接采集卡存在，内存即使自己回收，也还是会有问题



---Original---
From: ***@***.***&gt;
Date: Thu, Oct 12, 2023 19:47 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [secile/UsbCamera] Memory issues (Issue #27)




  
改用其他库，该库存在明显的内存回收问题，起因是外接的usb设备无法管理
  
并没有问题哦  我有项目跑了2年了 前台登记的客户端 一点内存都没漏
 只是例子里  没释放pictureBox1.Image旧图片
 
你用啥库 只要是把图片clone一份到pictureBox1.Image 都得自己释放

**Comment by yangjieshao** [comment]
> 您的库在系统自带的摄像头下无任何问题，一旦有其他的外接采集卡存在，内存即使自己回收，也还是会有问题
> […](#)
> ---Original--- From: ***@***.***&gt; Date: Thu, Oct 12, 2023 19:47 PM To: ***@***.***&gt;; Cc: ***@***.******@***.***&gt;; Subject: Re: [secile/UsbCamera] Memory issues (Issue #27) 改用其他库，该库存在明显的内存回收问题，起因是外接的usb设备无法管理 并没有问题哦 我有项目跑了2年了 前台登记的客户端 一点内存都没漏 只是例子里 没释放pictureBox1.Image旧图片 你用啥库 只要是把图片clone一份到pictureBox1.Image 都得自己释放 — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

你得翻译成英文啊 作者看不懂中文
然后 USB摄像头我跑了很久没问题
采集卡的话 我没有 不知道什么情况

现在的项目都要考虑信创 不用这个库了

**Comment by SankarGaneshKumar** [comment]
Tried your change still it leads to out of memory error. Only one time GC is calling atr 3.5 GB. then it leads to out of memory exception. When I changed this function 

 public int BufferCB(double SampleTime, IntPtr pBuffer, int BufferLen)
            {
                 if (Buffer == null || Buffer.Length != BufferLen)
                 {
                     Buffer = new byte[BufferLen];
                 }
                 // replace lock statement to Monitor.TryEnter. (issue #14)
                 var locked = false;
                 try
                 {
                     System.Threading.Monitor.TryEnter(BufferLock, 0, ref locked);
                     if (locked)
                     {
                         Marshal.Copy(pBuffer, Buffer, 0, BufferLen);
                       
                     }
                 }
                 finally
                 {
                     if (locked) System.Threading.Monitor.Exit(BufferLock);
                    Buffer = null;
                }
                 // notify buffered to worker thread. (issue #16)
                 if (Buffered != null) BufferedEvent.Set();
                return 0;
            }
            
            To this
            
            
             public int BufferCB(double SampleTime, IntPtr pBuffer, int BufferLen)
            {
                
                byte[] tempBuffer = new byte[BufferLen];
                
                var locked = false;
                try
                {
                    System.Threading.Monitor.TryEnter(BufferLock, 0, ref locked);
                    if (locked)
                    {

                        Marshal.Copy(pBuffer, tempBuffer, 0, BufferLen);
                        //Console.WriteLine("BufferLen " + BufferLen);
                        Buffer = tempBuffer;
                        tempBuffer = null;
                    }
                }
                finally
                {
                    if (locked) System.Threading.Monitor.Exit(BufferLock);
                }
                // notify buffered to worker thread. (issue #16)
                if (Buffered != null) BufferedEvent.Set();
                return 0;
            }
            
            GC is calling very frequently. But it is not leading to Out of memory exception. Any idea what is happening ?
![Screenshot (2)](https://github.com/secile/UsbCamera/assets/18550832/a392a2af-bea2-4f8a-a848-764faab6b982)

**Comment by yangjieshao** [comment]
@SankarGaneshKumar   maybe  you can try [FlashCap](https://github.com/kekyo/FlashCap)

It can run on Linux and Windows
if you wang run on mac  you can try  https://github.com/kekyo/FlashCap/pull/65

**Comment by secile** [comment]
hello, SankarGaneshKumar.

sorry, I do not have time, so in short.

I have ever experienced this problem once.
At that time, when I changed target framework .NET Framework version, the problem is disapeared.
Could you try?

**Comment by lwoldt** [comment]
我已经重写该部分 感谢你的帮助



---Original---
From: ***@***.***&gt;
Date: Thu, Oct 12, 2023 22:52 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [secile/UsbCamera] Memory issues (Issue #27)




 
hello, SankarGaneshKumar.
 
sorry, I do not have time, so in short.
 
I have ever experienced this problem once.
 At that time, when I changed target framework .NET Framework version, the problem is disapeared.
 Could you try?

**Comment by secile** [comment]
If it doesn't work, there's one more thing you can try.
Recently, I added 'USBCAMERA_BYTEARRAY' symbol.
if 'USBCAMERA_BYTEARRAY' symbol defined, GetBitmap() returns image data as byte array.

Try the code below.

```cs
// 2. use Timer and GetBitmap().
var timer = new System.Timers.Timer(1000 / 30) { SynchronizingObject = this };
timer.Elapsed += (s, ev) =>
{
    var buf = (byte[])camera.GetBitmap();
    pictureBox1.Image = BufferToBitmap(buf, camera.Size.Width, camera.Size.Height);
};
timer.Start();

private static Bitmap BufferToBitmap(byte[] buffer, int width, int height)
{
    var result = new Bitmap(width, height);
    var bmpData = result.LockBits(new Rectangle(Point.Empty, result.Size), System.Drawing.Imaging.ImageLockMode.WriteOnly, System.Drawing.Imaging.PixelFormat.Format24bppRgb);
    System.Runtime.InteropServices.Marshal.Copy(buffer, 0, bmpData.Scan0, buffer.Length);
    result.UnlockBits(bmpData);
    return result;
}
```

**Comment by secile** [comment]
I would like to summarize what we currently know about this issue.

* If you use a timer (or subscribe PreviewCaptured) instead of SetPreviewControl, an error may occur.
	* in my environment, when VideoFormat image size is greater that 1920×1080, error happens.
	* when VideoFormat image size is smaller than 1280×1024, error not happen.
* It is thought that the reason why error happen is out of memory due to Bitmap not being disposed.
* I previously said 'changed target framework .NET Framework version, the problem is disapeared', but maybe this is not correct.
* I previously said ''define USBCAMERA_BYTEARRAY' symbol', this has an effect.
	* the reason may that creating new Bitmap in local function make bitmap GC generation to 0, therefor, bitmap is disposed aggressively.

Basically, bitmap object is disposed automatically by GC, so it is not occur out of memory error in normal situation.
But for various reasons, for example, selecting large image size, bitmap is not disposed collectly and error occur.
If you are in the situation, you have to dispose bitmap expressly.
Please see also [issue34](https://github.com/secile/UsbCamera/issues/34).

I'll remain this issue open for a while.
If the same problem is not reported after a while, I will close this issue.

**Comment by zhangsan12138222** [comment]
> @SankarGaneshKumar maybe you can try [FlashCap](https://github.com/kekyo/FlashCap)
> 
> It can run on Linux and Windows if you want run on mac you can try [kekyo/FlashCap#65](https://github.com/kekyo/FlashCap/pull/65)

请问为什么我电脑休眠之前已经释放资源了，唤醒以后再打开摄像头会卡死。
